lib/Platform/Unicode/CharacterProperties.cpp

```diff
@@ -210,4 +211,148 @@ uint32_t canonicalize(uint32_t cp, bool unicode) {
+      begin, end, poolRef, [](const T &a, const StringPoolRef &b) {
+        return UNICODE_DATA_STRING_POOL.compare(
+                   a.name.offset,
+                   a.name.size,
+                   UNICODE_DATA_STRING_POOL,
+                   b.offset,
+                   b.size) < 0;
+      });
+  if (it == end ||
+      (it->name.offset != poolRef.offset && it->name.size != poolRef.size)) {
+    return end;
+  }
+  return it;
+}
+
+bool addUnicodePropertyRanges(
+    CodePointSet *receiver,
+    const std::string_view &propertyName,
+    const std::string_view &propertyValue,
+    bool inverted = false) {
+  auto key = propertyName;
+  auto rangeMapStart = std::begin(unicodePropertyRangeMap_BinaryProperty);
+  auto rangeMapEnd = std::end(unicodePropertyRangeMap_BinaryProperty);
+  auto nameMapStart = std::begin(canonicalPropertyNameMap_BinaryProperty);
+  auto nameMapEnd = std::end(canonicalPropertyNameMap_BinaryProperty);
+
+  if (propertyValue.empty()) {
+    // There was no property value, this is either a binary property or a value
+    // from General_Category, as per `LoneUnicodePropertyNameOrValue`.
+    if (findNameMapEntry(nameMapStart, nameMapEnd, key) == nameMapEnd) {
+      rangeMapStart = std::begin(unicodePropertyRangeMap_GeneralCategory);
+      rangeMapEnd = std::end(unicodePropertyRangeMap_GeneralCategory);
+      nameMapStart = std::begin(canonicalPropertyNameMap_GeneralCategory);
+      nameMapEnd = std::end(canonicalPropertyNameMap_GeneralCategory);
+    }
+  } else {
+    // There was a property value, assume the name is a category.
+    key = propertyValue;
+    if (propertyName == "General_Category" || propertyName == "gc") {
+      rangeMapStart = std::begin(unicodePropertyRangeMap_GeneralCategory);
+      rangeMapEnd = std::end(unicodePropertyRangeMap_GeneralCategory);
+      nameMapStart = std::begin(canonicalPropertyNameMap_GeneralCategory);
+      nameMapEnd = std::end(canonicalPropertyNameMap_GeneralCategory);
+    } else if (propertyName == "Script" || propertyName == "sc") {
+      rangeMapStart = std::begin(unicodePropertyRangeMap_Script);
+      rangeMapEnd = std::end(unicodePropertyRangeMap_Script);
+      nameMapStart = std::begin(canonicalPropertyNameMap_Script);
+      nameMapEnd = std::end(canonicalPropertyNameMap_Script);
+    } else if (propertyName == "Script_Extensions" || propertyName == "scx") {
+      // Script_Extensions is a superset of Script.
+      if (!addUnicodePropertyRanges(receiver, "Script", propertyValue)) {
+        return false;
+      }
+      rangeMapStart = std::begin(unicodePropertyRangeMap_ScriptExtensions);
+      rangeMapEnd = std::end(unicodePropertyRangeMap_ScriptExtensions);
+      nameMapStart = std::begin(canonicalPropertyNameMap_Script);
+      nameMapEnd = std::end(canonicalPropertyNameMap_Script);
+    } else {
+      return false;
+    }
+  }
+
+  // Canonicalize the property name.
+  auto canonicalNameEntry = findNameMapEntry(nameMapStart, nameMapEnd, key);
```

Comment:
I'm finding it a little hard to follow the naming here between `name`, `value`, and `key`. Are there more precise terms we can use? Otherwise, a comment describing what each thing refers to would be helpful.

Reply:
I can't say I blame you, I think it's confusing too and would really appreciate finding some better names. Maybe I can explain the terms, and you can decide if there are better names or if perhaps a comment is more useful.

The RegExp Unicode property syntax accepts:
- Binary properties, where the presence of the property implies a value of true, e.g. `ASCII_Hex` , `Emoji`, etc.
   - Additionally, all of `General_Category` is also allowed to be specified as a binary property, e.g. `Lowercase_Letter`, `N`, `Symbol`, etc.
- Properties with a value, where the value is not a binary choice, e.g. `Scripts=Latin`, `Script_Extensions=Katakana`, `General_Category=Symbol`, etc.

The "property name" and "property value" terminology comes from [ES262](https://tc39.es/ecma262/multipage/text-processing.html#prod-UnicodePropertyValueExpression).

The "key" terminology is my own, and is imprecise because it is used to resolve a name to its canonical form, by looking it up in `canonicalPropertyNameMap_<CATEGORY>`, where `CATEGORY` is determined as follows:

- If there is no `propertyValue` (in the `PropertyName=PropertyValue` sense) given, then `CATEGORY` assumed to be `BinaryProperty`
  - If that assumption is incorrect, and it's not in the binary property category, then try `GeneralCategory` 
- If there is a `propertyValue`, then `CATEGORY` is determined by `propertyName` (e.g. `canonicalPropertyNameMap_Script`)

In both cases `key` needs to be resolved according to the canonical name of the _value_, but in the binary property case the "value" can be thought of something like `General_Category=VALUE`. I chose `key` in the sense of `value = dict[key]` but maybe a term like `needle` better conveys the "we need to find this" idea.

Reply:
Makes sense. Perhaps the best thing to do is to compute `canonicalNameEntry ` in each of the cases, so the logic is easier to follow. It would also avoid duplicating the lookup for binary properties. So for example:

```
  NameMapEntry *canonicalNameEntry;
  llvh::ArrayRef<RangeMapEntry> rangeMap;

  if (propertyValue.empty()) {
    // There was no property value, this is either a binary property or a value
    // from General_Category, as per `LoneUnicodePropertyNameOrValue`.
    if(canonicalNameEntry = findNameMapEntry(canonicalPropertyNameMap_BinaryProperty, propertyName))
      rangeMap = unicodePropertyRangeMap_BinaryProperty;
    else if(canonicalNameEntry = findNameMapEntry(canonicalPropertyNameMap_GeneralCategory, propertyName))
      rangeMap = unicodePropertyRangeMap_GeneralCategory;
  }
...
```

Reply:
That's a great suggestion, thanks! It definitely helps localise the initial setup, before moving onto the range lookups.

I did have to do `findMapEntry(llvh::ArrayRef(…), …);` to appease compiler, let me know if you have any thoughts on that.